src/lseq/ident.rs

```diff
@@ -0,0 +1,409 @@
+    //
+    // This reflects the case where we want to allocate something between 0.2 and 0.200000001
+    fn alloc_from_upper(
+        &mut self,
+        base: &Identifier<A>,
+        q: &Identifier<A>,
+        mut depth: usize,
+    ) -> Identifier<A> {
+        let mut ident = base.clone();
+        loop {
+            match q.path.get(depth) {
+                // append a 0 to the result path as well
+                Some(b) if b.0 <= 1 => ident.path.push((0, b.1.clone())),
+                // oo! a non-zero value
+                _ => break,
+            }
+            depth += 1;
+        }
+
+        // If we actually ran out of upper bound values then we're free to choose
+        // anything on the next level, otherwise use the upper bound we've found.
+        let upper = match q.path.get(depth) {
+            Some(b) => b.0,
+            None => self.width_at(depth as u32 + 1),
+        };
+        let next_index = self.index_in_range(1, upper, depth as u32);
+        self.push_index(&ident, next_index)
+    }
+
+    fn replace_last(&mut self, p: &Identifier<A>, depth: usize, ix: u64) -> Identifier<A> {
+        let mut ident = p.clone();
+        ident.path.truncate(depth);
+        ident.path.push((ix, Some(self.site_id.clone())));
+        ident
+    }
+
+    fn push_index(&mut self, p: &Identifier<A>, ix: u64) -> Identifier<A> {
+        let mut ident = p.clone();
+        ident.path.push((ix, Some(self.site_id.clone())));
+        ident
+    }
+
+    fn width_at(&self, depth: u32) -> u64 {
+        assert!(self.initial_base_bits + depth  < 31, "maximum depth exceeded");
+
+        2u64.pow(self.initial_base_bits + depth)
+    }
+    // Generate an index in a given range at the specified depth.
+    // Uses the allocation strategy of that depth, boundary+ or boundary- which is biased to the
+    // lower and upper ends of the range respectively.
+    // should allocate in the range [lower, upper)
+    fn index_in_range(&mut self, lower: u64, upper: u64, depth: u32) -> u64 {
+        assert!(
+            lower < upper,
+            "need at least one space between the bounds lower={} upper={}",
+            lower,
+            upper
+        );
+
+        let mut rng = rand::rngs::OsRng;
+        let interval = BOUNDARY.min(upper - 1 - lower);
+        let step = if interval > 0 {
+            rng.gen_range(0, interval)
+        } else {
+            0
+        };
+        if self.strategy(depth) {
+            //boundary+
+            lower + step
+        } else {
+            //boundary-
+            upper - step - 1
+        }
+    }
+
+    fn strategy(&mut self, depth: u32) -> bool {
+        match self.strategy_vec.get(depth as usize) {
+            None => {
+                self.strategy_vec.push(rand::thread_rng().gen());
+                self.strategy(depth)
+            }
+            Some(s) => *s,
+        }
+        // temp strategy. Should be a random choice at each level
+    }
+}
+
+#[cfg(test)]
```

Comment:
```suggestion
```
Don't constrain the arbitrary impls to tests, It's useful to have access to these in user code if they want to do some property testing

Reply:
Superseded by https://github.com/xldenis/rust-crdt/pull/6